Foundation for v2 serialization and deserialization

bshepherdson opened the pull request:
This PR lays the foundation for the new serdes. So far it only applies to collections and settings, because I want to get feedback on and evolve the design before investing in all the models.

Please take particular care over the factoring in `metabase.models.serialization.base`, and how the various steps of the process fit together.

bshepherdson committed:
Starting point for serialization v2

This supports serialization of only Collections and Settings so far, but
it demonstrates the design of the new serialization system.

`metabase.models.serialization.base` has the protocols involved, which
are to be implemented by all the exported models eventually.
The actual serialization code that drives the larger process is in
`metabase_enterprise.serialization.serialize` and `.deserialize`, since
serialization is an enterprise feature.

bshepherdson commented:
```clojure
(into [] (serdes.utils/serdes-serialize-all Collection 1))

[["Collection/s5RLDUrpcPWFpmsGSpkr_.yaml"
  {:authority_level nil,
   :description nil,
   :archived false,
   :slug "some_collection",
   :color "#509EE3",
   :name "Some collection",
   :personal_owner_id nil,
   :serdes-type "Collection",
   :entity_id "s5RLDUrpcPWFpmsGSpkr_",
   :namespace nil}]
 ["Collection/_sm8wK8dEfwxJRwYA32EY.yaml"
  {:authority_level nil,
   :description nil,
   :archived false,
   :slug "nested_collection",
   :color "#509EE3",
   :name "Nested collection",
   :personal_owner_id nil,
   :parent_id "s5RLDUrpcPWFpmsGSpkr_",
   :serdes-type "Collection",
   :entity_id "_sm8wK8dEfwxJRwYA32EY",
   :namespace nil}]
 ["Collection/4ZqBc7pg91cEGan1bqffl.yaml"
  {:authority_level nil,
   :description nil,
   :archived false,
   :slug "braden_shepherdson_s_personal_collection",
   :color "#31698A",
   :name "Braden Shepherdson's Personal Collection",
   :personal_owner_id 1,
   :serdes-type "Collection",
   :entity_id "4ZqBc7pg91cEGan1bqffl",
   :namespace nil}]]
```

qnkhuat commented on src/metabase/models/serialization/utils.clj:
```diff
@@ -65,3 +68,152 @@
+  [])
+
+(p.types/defprotocol+ ISerializable
+  ;;; Serialization side
```
what do you think if we prefix all methods for serialization with `ser` or `serialize`, and for deserialization : `des` or `deserialize` ?

qnkhuat commented on src/metabase/models/serialization/utils.clj:
```diff
@@ -65,3 +68,152 @@
+  (db/update! old-entity (get old-entity (models/primary-key old-entity)) new-map))
+
+(defn- default-insert [model new-map]
+  (db/insert! model new-map))
```
I wonder whether the toucans hook will affect these insert? Does these insert require all hooks to be `idempotent`?

For example in User/pre-insert hook, we hash the password. So does that mean if we deserialize it here, the inserted user will have password is the hash of a `hashed` password?

What could go wrong if we do `simple_insert`? which will by pass all hooks

qnkhuat replied:
same qs for `db/update!`

bshepherdson replied:
We almost certainly want `simple-insert`, actually. We don't want, for example, the import to generate `entity_id`s. I hadn't thought about that.

Is there a similar concern for an update?

bshepherdson committed:
Create `metabase-enterprise.serialization.v2` and move serdes code there

Only the parts that needs to be implemented for each model need to be in
`metabase.models.serialization.base`.

bshepherdson committed:
Some real tests for serialize

bshepherdson committed:
Test the fundamentals of the serialization system

bshepherdson committed:
Start calling Collections' files Collection/entity_id+slug.yaml

The slug is a ready-made filesystem-friendly field, and it makes the
Collection directory much more navigable by humans.

bshepherdson committed:
Add deserialization tests for a few permutations

Including testing entity_id de-duplication and identity-hashes.

bshepherdson committed:
lint

bshepherdson committed:
lint

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/v2/serialize.clj:
```diff
@@ -0,0 +1,14 @@
+(ns metabase-enterprise.serialization.v2.serialize
+  "Top-level functionality for serializing a Metabase appdb into YAML files for import into another.
```
I would suggest separating out the task of `stuff in app DB -> YAML` into two separate tasks

1. `stuff in app DB -> serializable form(s)` (e.g. generic Clojure data structures)
2. serializable form(s) -> YAML

(and vice versa for deserialization).

That would make things a lot easier to test I think and keep concerns separate. Also not saying we will want to support any formats other than YAML but serialization is not inherently a YAML thing so it seems like tying it in to YAML is unnecessary coupling

bshepherdson replied:
That's actually how it's designed; there's no YAML anywhere except the filename suffixes. That could be dropped from the filenames so that any format would work.

bshepherdson replied:
Done.

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/v2/serialize.clj:
```diff
@@ -0,0 +1,14 @@
+            [metabase.models.serialization.base :as serdes.base]
+            [toucan.db :as db]))
+
+(defn serialize-metabase
+  "Serializes the complete database into a reducible stream of [file-path edn-map] pairs.
+  This is the last step before conversion to YAML and writing to disk, and a useful point for testing.
+  The file paths are relative to the root dump directory."
+  [user-or-nil]
```
It's not really clear to me from this docstring what the purpose of `user-or-nil` is

bshepherdson replied:
This is now formulated as an options map, and properly documented.

camsaul commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,171 @@
+        (deserialize-upsert (db/select-one model :id local) hydrated)
+        (deserialize-insert model hydrated)))))
+
+(def ISerializableDefaults
+  "Default implementations for [[ISerializable]], so models need only override those that need special handling."
+  {:serialize-all         default-serialize-all
+   :serialize-query       default-query
+   :serialize-one         (serialize-one-plus identity nil)
+   :deserialize-file      (deserialize-file-plus identity)
+   :deserialize-upsert    default-upsert
+   :deserialize-insert    default-insert})
```
I would vote making these multimethods instead. Then you don't need to muck around with `ISerializableDefaults` hacks and stuff like that. It's more idiomatic too.

(I know Toucan does things this way but it's a dumb way of doing things and I wouldn't do it that way again. Toucan 2 uses multimethods for stuff like `pre-insert` and `post-select`.)

bshepherdson replied:
I've made it so with a major refactoring. I like the way the multimethods work!

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/v2/deserialize.clj:
```diff
@@ -0,0 +1,79 @@
+
+(defn- path-parts [path]
+  (->> (java.nio.file.Paths/get path (into-array String []))
+       (.iterator)
+       (iterator-seq)
```
```suggestion
       .iterator
       iterator-seq
```

If you don't pass any args the `()` is optional and included automatically in the threading macros

bshepherdson replied:
This code is on ice. Duly noted for future style.

camsaul commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,171 @@
+
+(defn- default-upsert [old-entity new-map]
+  (db/update! (symbol (name old-entity)) (get old-entity (models/primary-key old-entity)) new-map)
+  #_(prn "upsert" old-entity new-map))
```
I would change these to `log/trace` or `log/debug` and uncomment them so you can see this stuff while working on it.

```clj
(metabase.test/set-ns-log-level! 'metabase.models.serialization.base :trace)
```

is a tool I wrote a while back that will let you flip the log level of a specific namespace from the REPL.

bshepherdson replied:
Oh, cool tool.

Those are actually leftover debugging code from when I didn't want to change my dev appdb. I'll turn them into permanent `log/debug`s.

camsaul commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,171 @@
+    entity.
+    Defaults to a naive update by primary key, using just the new map's values.")
+
+  (deserialize-insert
```
Anything method or function that modifies global state (in this case, the app DB) should end in a `!`. So this should probably be called `deserialize-insert!`

camsaul replied:
Same goes for the other functions with side effects

bshepherdson replied:
Done, here and elsewhere.

camsaul commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,171 @@
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
+    "Given the model, the file contents as an EDN map, and any corresponding entity already existing in the database,
+    this performs whatever steps are necessary to update or insert the incoming entity.
+    Called for each **file**, which usually means for each entity. For single-file output like for settings, this is
+    called once for the whole file.
+    Defaults to [[deserialize-file-plus]] with [[identity]] as the inner transformation.")
```
I like the fact that this is decoupled from the actual format of the files themselves (i.e. YAML) and the implementation doesn't have to worry about that. I suggest taking the abstraction a bit further tho. It seems a little weird that a model's implementation has to concern itself with whether or not some specific file contained a single object or multiple objects. It seems like that's some thing that can be handled in the shared code and the actual implementations could be `deserialize-one` or `deserialize-many` instead. `deserialize-many` could just repeatedly call `deserialize-one` by default

bshepherdson replied:
I've made the refactoring you suggested; the main deserialization now has no connection to files at all, just an abstracted `ingestion` surface.

camsaul commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,171 @@
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
```
It would be SUPER DUPER helpful to document every variable in methods like this that we expect to implement in a few dozen places. I do not know what `local` means here. I'm guessing  it's a boolean value (in which case I would personally name it `local?` but that's a style preference hill that I'm not willing to die on)? 

The only way to figure out what `local` means/is used for is to go read implementations of the method, which is definitely a sign the docstring isn't detailed enough. Ideally we should know what a function does, what it returns, and what all the parameters should be/are used for just by reading the docstring.

bshepherdson replied:
Docs are a lot better in the rebuilt PR.

camsaul commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,171 @@
+  "Given a function, this returns a function suitable for use as [[deserialize-file]]. The provided function is applied
+  to the EDN map read from the YAML file, and then either [[deserialize-upsert]] or [[deserialize-insert]] is called
+  depending on whether a corresponding entity was found in the database."
+  [f]
```
A few thoughts: 

- I would name `f` something like `xform` or `contents-xform` or better yet `contents-transform-fn` to make it clearer that it's a function used to transform the file contents. Personally I am of the you-can't-have-variable-names-that-are-too-long school. I think the longer names make the code easier to read
- Since this is a function that given a transform function returns a `deserialize-file` function I would name this function something like `deserialize-file-fn` or `deserialize-file-with-transform-fn`. Sticking `-fn` on the end makes it clear that it returns a `-fn`. I don't know where you got `-plus` from but NGL it confused me a lot when I first saw it

bshepherdson replied:
The new factoring is both more transparent and better documented.

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/v2/deserialize.clj:
```diff
@@ -0,0 +1,79 @@
+
+(defn deserialize-metabase
+  "Deserializes a complete database export from a 'deserialization source', which can be created with
+  `deserialization-source-foo`. This doesn't read directly from files for ease of testing."
```
`deserialization-source-foo` 

huh?

bshepherdson replied:
This lame naming convention has been replaced with an `Ingestable` protocol, which is properly documented.

camsaul reviewed (dismissed):
Overall this looks great. It's already 100x easier to follow then the old code and looks like it's tested a lot better too. I have a few small requests however

1. I would use multimethods here instead of a protocol. Less of a hassle and more idiomatic
2. I know in practice we're serializing/deserializing from YAML but I think the code would benefit from you pretending that we don't know what the file format is going to be (or even that we're writing to a file at all) everywhere except for the tiny bit of code that actually needs to read and write to those files. Decouple the YAML-ness from every other part of the code. That will make things easier to test and work with. You should do this in a way that keeps everything reducible tho. e.g Part one will take a reducible sequence of objects of some type and return reducible "serializable" plain-Clojure-map versions (using an `eduction` or reifying an `IReduceInit` or something like that) and Part two will take the reducible sequence of plain Clojure maps and write them to a YAML file or files.

   On that note I think you might want to tweak methods like `deserialize-file` so they're independent of the notion of things coming from a file. Maybe something like `deserialize-many` and/or `deserialize-one` would work better here. Then the individual model implementations don't have to worry about the format of the serialization files themselves or whatever.
   
   Instead of asking `serialize-one` or whoever does it to come up with their own `.yaml` filename have them come up with a globally-unique `:serder/id` or something and have the plain map -> YAML code take that ID and turn it into a filename.
4. Please give methods and functions that have side effects (such as inserting or updating rows in the application DB) names that end in `!`
5. I left a few small style notes about suggested names for stuff and places where parameters weren't mentioned in docstrings -- since these methods are going to be implemented all over the place they should be really well documented
6. Hopefully it's not too early to start thinking about this but I would seriously consider trying to write a generative test suite as you go with this project, and as you add support for new models, add them to the generative suite as well. It will cover a lot more things than just writing a handful of specific tests. And it will be a lot easier to do it a few models at a time instead of all at once. For example here you can take all of our various settings and give them random values based on the Setting `:type` and generate some Collections maybe with parent/grandparent/etc. Collections and various different properties. Then you just take the randomly generated data and serialize it and deserialize it and make sure the properties are the same. This is not a blocker for this PR, but it is something that the sooner we start as part of the refactor, the better IMO

bshepherdson commented:
> Overall this looks great. It's already 100x easier to follow then the old code and looks like it's tested a lot better too. I have a few small requests however
> 
> 1. I would use multimethods here instead of a protocol. Less of a hassle and more idiomatic

Will do. I went back and forth on this a couple of times and ended up following Toucan, as you surmised.

> 2. I know in practice we're serializing/deserializing from YAML but I think the code would benefit from you pretending that we don't know what the file format is going to be (or even that we're writing to a file at all) everywhere except for the tiny bit of code that actually needs to read and write to those files. Decouple the YAML-ness from every other part of the code. That will make things easier to test and work with. You should do this in a way that keeps everything reducible tho. e.g Part one will take a reducible sequence of objects of some type and return reducible "serializable" plain-Clojure-map versions (using an `eduction` or reifying an `IReduceInit` or something like that) and Part two will take the reducible sequence of plain Clojure maps and write them to a YAML file or files.
>    On that note I think you might want to tweak methods like `deserialize-file` so they're independent of the notion of things coming from a file. Maybe something like `deserialize-many` and/or `deserialize-one` would work better here. Then the individual model implementations don't have to worry about the format of the serialization files themselves or whatever.
>    Instead of asking `serialize-one` or whoever does it to come up with their own `.yaml` filename have them come up with a globally-unique `:serder/id` or something and have the plain map -> YAML code take that ID and turn it into a filename.

I like the idea of a unique `:serdes/id` that can be used as a filename. Actually, since I already put a `:serdes_type` on it, perhaps putting `{:serdes/info {:type "Collection" :id "1234567890abcdefghijkl" :label "human readable name"} ...}` on everything?

> 3. Please give methods and functions that have side effects (such as inserting or updating rows in the application DB) names that end in `!`

Will do.

> 4. I left a few small style notes about suggested names for stuff and places where parameters weren't mentioned in docstrings -- since these methods are going to be implemented all over the place they should be really well documented

The quality got pretty patchy as I kept rewriting and refactoring parts of it. I'll polish them up.

> 5. Hopefully it's not too early to start thinking about this but I would seriously consider trying to write a generative test suite as you go with this project, and as you add support for new models, add them to the generative suite as well. It will cover a lot more things than just writing a handful of specific tests. And it will be a lot easier to do it a few models at a time instead of all at once. For example here you can take all of our various settings and give them random values based on the Setting `:type` and generate some Collections maybe with parent/grandparent/etc. Collections and various different properties. Then you just take the randomly generated data and serialize it and deserialize it and make sure the properties are the same. This is not a blocker for this PR, but it is something that the sooner we start as part of the refactor, the better IMO

Probably not part of this PR, but I've already called out generative testing in the [epic](https://github.com/metabase/metabase/issues/22641). Serialization is a great use case for it.

bshepherdson committed:
Major overhaul to multimethods, somewhat different factoring

Now in four phases:
serialization = extract + store;
deserialization = ingest + merge.

File names are now a purely store + ingest concern; extract and merge
deal use a `:serdes/meta {:id "entity ID" :type "Collection"}` convention
and so are storage agnostic.

Laziness is retained: merging ingests full details on demand, so only
the metadata of the importing database needs to fit in memory.

bshepherdson commented:
PTAL! I've landed a big refactor following Cam's suggestions. Here's an overview of the new design; see `metabase.models.serialization.base` for a fuller detail:

- The exportable form in memory is a vanilla Clojure map with a `:serdes/meta {:id "1234567..." :type "Collection"}` key-value; this is the only convention shared by the decoupled parts of the process.
- Serialization is decoupled into `extract` plus `store`.
    - Extraction produces a reducible stream of exportable maps;
    - Storing actually puts them in files or on the network.
- Deserialization is decoupled into `ingest` plus `merge`.
    - Ingestion reads the disk (or network) format and produces a list of metadata, or the full details of a specific entity.
    - Merging reduces over that list, fetching each model's details and merging it in.
- The heavy lifting is done by multimethods in `metabase.models.serialization.base`, with overrides in eg. `metabase.models.collection`.
- The top-level processes are in `metabase-enterprise.serialization.v2.*`

There are tests for extraction and merging, with a dummy in-memory `Ingestable`. Later PRs will add the YAML conversion as a nicely decoupled, testable store + ingest pair. It can do custom things like combining Settings into one file without the `extract` or `merge` code having to know.

If you have better names for these four stages to suggest, I'm listening. I'm not very happy with "merge". I deliberately didn't use "(de)serialize" for the stages; I want to reserve those for the whole process.

qnkhuat commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,324 @@
+;;; This two-stage design avoids needing all the data in memory at once, where that's practical with the underlying
+;;; some storage media (eg. files).
+;;;
+;;; Merging:
```
How about Load? as in Extract Transform Load

qnkhuat commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,324 @@
+  Returns the primary key of the updated entity."
+  (fn [model _ _] model))
+
+(defmethod merge-upsert! :default [model ingested local]
```
why don't we call it update?

qnkhuat reviewed (commented):
I think we need to be careful about which toucan methods we use. 
We're using `db/select-reducible` in the extract phase and this method apply `post-select` hook and toucan/types to the result.

But we're using `db/simple-insert!` in the merge phase, and it doesn't run through toucan/types or hooks. This could result in mismatching data after Merge.

For example our Setting tables, the `value` column has a toucan type of [encrypted-text](https://github.com/metabase/metabase/blob/57a7cbfdbaf9d1bd09dc611d72bfc22bd4ea00c8/src/metabase/models/setting.clj#L142).

This type will automatically encrypt/decrypt if a `MB_ENCRYPTION_SECRET_KEY` key is set.

So during the Extract phase, we use get settings by `db/select-reducible` and these values will be decrypted.

Then in the Merge phase, we use `db/simple-insert!` and it'll insert raw Settings into DB.

In this particular case, it will not result in breaking Metabase because the `encrypted-text` type has some magic to not decrypt if it's a raw value.
But the problem remains, that the mismatched in `extract` and `merge` could potentially result in corrupted data. I'm not sure how this will work for types like `:json` or `:keyword` though.

FWIW, I think it's best if we bypass all types and hooks from toucan in both the extract and merge phase. That's way serdes will basically deal with raw value in and out of appDB.

bshepherdson committed:
Move away from high-level Toucan functions to lower level JDBC

Though it's actually still calling through `toucan.db/reducible-query`
and other public helpers. Toucan is still useful for its default DB
connection, metadata (eg. what's the primary key) and handling the
quoting patterns properly.

Also changes the name of the deserialization phases to "ingest" +
"load" (formerly "merge").

qnkhuat commented on enterprise/backend/src/metabase_enterprise/serialization/v2/load.clj:
```diff
@@ -0,0 +1,92 @@
+    :else (let [ingested (serdes.ingest/ingest-one ingestion meta-map)
+                model    (db/resolve-model (symbol type))
+                deps     (serdes.base/serdes-dependencies ingested)
+                ctx'     (-> ctx
```
seems like we no longer need `ctx` after this point, so just use `ctx` instead of `ctx'` to avoid mistyping?

qnkhuat commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,367 @@
+     (serdes.hash/identity-hash (db/select-one model pk key)) ; TODO This sucks for identity-hash!
+     key]))
+
+(defn- ingested-type
```
```suggestion
(defn- ingested-model
```
?

bshepherdson replied:
Done.

qnkhuat commented on src/metabase/models/serialization/base.clj:
```diff
@@ -0,0 +1,367 @@
+        ;              #(.substring 5))
+        ]
+    (log/tracef "Upserting %s %d: old %s new %s" model-name id (pr-str local) (pr-str ingested))
+    ; Using the two-argument form of [[db/update!]] that takes the model and a HoneySQL form for the actual update.
```
oh, I didn't know that the 2-arity skip the types and hooks, I wonder if this was intentional?

Either way I think it's worth a more detailed comment saying that this is bc we want to skip all toucan's magics

qnkhuat reviewed (approved):
a couple of nits but this lgtm now.

Great document btw

bshepherdson committed:
review commentary

noahmoss reviewed (approved):
Nice, just some tiny nits (mostly in comments)

bshepherdson commented on enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj:
```diff
@@ -0,0 +1,71 @@
+(ns metabase-enterprise.serialization.v2.extract-test
+  (:require [clojure.test :refer :all]
+            [metabase-enterprise.serialization.test-util :as ts]
+            [metabase-enterprise.serialization.v2.extract :as sut]
```
Done, now `extract`.

bshepherdson committed:
A few small review fixes

bshepherdson committed:
fix an unrelated test
